### PR TITLE
Fix weak reference problem with Document and TextAndVersion

### PR DESCRIPTION
**Pull request by mattwar**
Prior to this change references to SourceText obtained from a Document were not keeping the weakly held text (cached, recoverable or not) alive.  This was due to Document implementation holding onto TextAndVersion instance instead of SourceText.

This change re-implements RecoverableTextAndVersion to store a recoverable SourceText.

RecoverableCachedObjectSource was renamed to RecoverableWeakValueSource and moved to utilities because it no longer is tied to caching.  

Added CachedWeakValueSource implementation (similar to RecoverableWeakValueSource, but the value is recovered from the initial value source instead of logic in derived class.

Moved some logic buried in ISyntaxTreeFactorService that caused CreateRecoverableSyntaxTree to return a different kind of tree into a separate method that can be queried before calling CreateRecoverableSyntaxTree because the caller needs to know which type is created because the kind text source supplied depends on the tree type being constructed.

### CODE REVIEW

**Comment by mattwar** [comment]
@rchande, @Pilchie, @pharring, @jasonmalinowski, @dpoeschl please review.

File: src/Workspaces/Core/Portable/Utilities/ValuesSources/CachedWeakValueSource.cs [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Text;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.Host
+{
+    /// <summary>
+    /// A value source that cache's its value weakly once obtained from its source.
```
Comment (pharring) [inline src/Workspaces/Core/Portable/Utilities/ValuesSources/CachedWeakValueSource.cs]: Nit: no apostrophe needed on "caches"

File: src/Workspaces/Core/Portable/Utilities/ValuesSources/RecoverableWeakValueSource.cs [hunk 2]
```diff
@@ -9,11 +9,13 @@
...
 {
     /// <summary>
-    /// This class is a ValueSource that manages storing an item in a cache,
-    /// saving the item's state when it is evicted from the cache and recovering
-    /// it from its saved state when the value source is accessed again.
+    /// This class is a <see cref="ValueSource{T}"/> that holds onto a value weakly, 
+    /// but can save its value and recover it on demand if needed.
+    /// 
+    /// The initial value comes from the <see cref="ValueSource{T}"/> specified in the constructor.
+    /// Derived types implement SaveAsync and RecoverAsync.
```
Comment (pharring) [inline src/Workspaces/Core/Portable/Utilities/ValuesSources/RecoverableWeakValueSource.cs]: nit: crefs for SaveAsync and RecoverAsync

**Comment by mattwar** [comment]
@dotnet-bot test this please

**Comment by mattwar** [comment]
@dotnet-bot test this please

File: src/Workspaces/Core/Portable/Workspace/Solution/RecoverableTextAndVersion.cs [hunk 3]
```diff
@@ -1,36 +1,65 @@
...
+            else
+            {
+                value = null;
+                return false;
+            }
+        }
+
         public bool TryGetTextVersion(out VersionStamp version)
         {
+<<<<<<< HEAD
```
Comment (pharring) [inline src/Workspaces/Core/Portable/Workspace/Solution/RecoverableTextAndVersion.cs]: Merge conflict

Comment (mattwar) [inline src/Workspaces/Core/Portable/Workspace/Solution/RecoverableTextAndVersion.cs]: maybe that's why its not building.

**Comment by Pilchie** [comment]
:+1:

**Comment by mattwar** [comment]
@dotnet-bot please test this

**Comment by mattwar** [comment]
@dotnet-bot test this please

File: src/Workspaces/Core/Portable/Workspace/Solution/DocumentState.cs [hunk 4]
```diff
@@ -423,37 +423,52 @@ private VersionStamp GetNewTreeVersionForUpdatedTree(SyntaxNode newRoot, Version
...
-                    : factory.CreateRecoverableTree(info.Id.ProjectId, GetSyntaxTreeFilePath(info), options, lazyText, encoding, newRoot),
-                treeVersion);
-
-            return Tuple.Create<ValueSource<TextAndVersion>, TreeAndVersion>(lazyText, lazyTree);
+            SyntaxTree tree = null;
+            ValueSource<TextAndVersion> lazyTextAndVersion = null;
+
+            if ((mode == PreservationMode.PreserveIdentity) || !factory.CanCreateRecoverableTree(newRoot))
+            {
+                // its okay to use a strong cached AsyncLazy here because the compiler layer SyntaxTree will also keep the text alive once its built.
```
Comment (pharring) [inline src/Workspaces/Core/Portable/Workspace/Solution/DocumentState.cs]: Nit: it's (with an apostrophe). "It's okay..." and "once it's built"

**Comment by pharring** [comment]
:+1:

File: src/Workspaces/CSharp/Portable/LanguageServices/CSharpSyntaxTreeFactoryService.cs [hunk 5]
```diff
@@ -48,8 +48,15 @@ public override SyntaxNode DeserializeNodeFrom(Stream stream, CancellationToken
                 return CSharpSyntaxNode.DeserializeFrom(stream, cancellationToken);
             }
 
+            public override bool CanCreateRecoverableTree(SyntaxNode root)
+            {
+                var cu = root as CompilationUnitSyntax;
+                return base.CanCreateRecoverableTree(root) && cu != null && cu.AttributeLists.Count == 0;
```
Comment (jasonmalinowski) [inline src/Workspaces/CSharp/Portable/LanguageServices/CSharpSyntaxTreeFactoryService.cs]: Why the special case around the compilation unit having attribute lists...?

Comment (jasonmalinowski) [inline src/Workspaces/CSharp/Portable/LanguageServices/CSharpSyntaxTreeFactoryService.cs]: Also, the cu != null check can probably happen before the base.CanCreateRecoverableTree(...) clause. Might be needed for correctness if a tree of a different language somehow flowed through here.

File: src/Workspaces/VisualBasic/Portable/LanguageServices/VisualBasicSyntaxTreeFactoryService.vb [hunk 6]
```diff
@@ -43,7 +43,13 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
                 Return SyntaxFactory.SyntaxTree(root, options, fileName, encoding)
             End Function
 
+            Public Overrides Function CanCreateRecoverableTree(root As SyntaxNode) As Boolean
+                Dim cu = TryCast(root, CompilationUnitSyntax)
+                Return MyBase.CanCreateRecoverableTree(root) AndAlso cu IsNot Nothing AndAlso cu.Attributes.Count = 0
```
Comment (jasonmalinowski) [inline src/Workspaces/VisualBasic/Portable/LanguageServices/VisualBasicSyntaxTreeFactoryService.vb]: Ditto comments from C# side.

File: src/Workspaces/CSharp/Portable/LanguageServices/CSharpSyntaxTreeFactoryService.cs (hunk 5 above)
Comment (mattwar) [inline src/Workspaces/CSharp/Portable/LanguageServices/CSharpSyntaxTreeFactoryService.cs]: The attribute list check is a currently a mystery.  I retained it when I refactored the code, but the only one that seems to have known why it is there is Kyle, as this change is one of the last things he did before he left.

**Pull request merged**